src/QueueProvider.php

```diff
@@ -0,0 +1,148 @@
+<?php
+declare(strict_types=1);
+
+namespace Yiisoft\Queue\Redis;
+
+use RedisException;
+use Yiisoft\Queue\Redis\Exception\NotConnectedRedisException;
+
+class QueueProvider implements QueueProviderInterface
+{
+    private const DEFAULT_CHANNEL_NAME = 'yii-queue';
+
+    /**
+     * @throws RedisException
+     */
+    public function __construct(
+        private \Redis $redis, //redis connection,
+        private string $channelName = self::DEFAULT_CHANNEL_NAME
+    )
+    {
+    }
+
+    /**
+     * @throws RedisException
+     */
+    public function pushMessage(string $message, array $metadata = []): int
+    {
+        $this->checkingConnect();
+        $id = $this->getId();
+        $this->redis->hset("$this->channelName.messages", (string) $id, $message);
+        $this->redis->lpush("$this->channelName.waiting", $id);
+        return $id;
+    }
+
+    /**
+     * @throws RedisException
+     */
+    public function existInWaiting(int $id): bool
+    {
+        $this->checkingConnect();
+        $exist = $this->redis->hexists("$this->channelName.messages", (string) $id);
+        return is_bool($exist) ? $exist : false;
+    }
+
+    /**
+     * @throws RedisException
+     */
+    public function existInReserved(int $id): bool
+    {
+        $this->checkingConnect();
+        $exist = $this->redis->hexists("$this->channelName.attempts", (string) $id);
+        return is_bool($exist) ? $exist : false;
+    }
+
+    /**
+     * @throws RedisException
+     */
+    public function reserve(int $timeout = 0): ?Reserve
+    {
+        $this->checkingConnect();
+        // Moves delayed and reserved jobs into waiting list with lock for one second
+        try {
+            if ($this->redis->set("$this->channelName.moving_lock", 'true', ['NX', 'EX', 1])) {
+                $this->moveExpired("$this->channelName.delayed");
+                $this->moveExpired("$this->channelName.reserved");
+            }
+        } finally {
+            $this->redis->del("$this->channelName.moving_lock");
+        }
+
+        $result = $this->redis->brpop("$this->channelName.waiting", $timeout);
+        if (is_null($result) || !isset($result[1])) {
+            return null;
+        }
+        $id = $result[1];
+        if (!is_string($id)) {
+            return null;
+        }
+
+        $payload = $this->redis->hget("$this->channelName.messages", $id);
+        if (!is_string($payload)) {
+            return null;
+        }
+        $this->redis->zRem("$this->channelName.reserved", time(), $id);
+        $this->redis->hincrby("$this->channelName.attempts", $id, 1);
+        return new Reserve((int) $id, $payload);
+    }
+
+    /**
+     * @throws RedisException
+     */
+    public function delete(string $id): void
+    {
+        $this->checkingConnect();
```

Comment:
```suggestion
        $this->checkConnection();
```

Also, we can connect automatically instead of throwing an exception.

Reply:
The client redis requires parameters (host, port etc) in the method connect.
```
$redis = new Redis();
//Connecting to Redis
$redis->connect('hostname', port);
```
I doubt that it makes sense to receive and store these parameters within this class.